### ISSUE TITLE
install.sh hangs indefinitely due to grep with one argument

### ISSUE THREAD

**Issue by ryantate** [freedomofpress/securedrop#1394]
I hit this issue while running `install.sh` on a fresh new TAILS stick, WITHOUT the ATHS info, so entering `torrc` lines via the GUI text prompts.

Everything went fine until eventually — right between `set_permissions_on_securedrop_dotfiles` and `create_desktop_shortcuts` in `main` — install.sh would just sit there seemingly stuck. I did some rudimentary print (err, `echo`) debugging in the shell script and discovered it would hang here, on line 222 of `install.sh`:

`app_document_aths="$(grep ^Exec=/usr/local/bin/tor-browser | awk '{ print $2 }')"`
https://github.com/freedomofpress/securedrop/blob/develop/tails_files/install.sh#L222

The issue seems to be this:

`grep ^Exec=/usr/local/bin/tor-browser`

This is `grep` with one argument, which causes `grep` to hang waiting for STDIN. Looking at the context of the shell script it's pretty clear there should be a second token here specifying which file to grep, probably `"${amnesia_desktop}/document.desktop"` checked for on line 221.

**Comment by fowlslegs** [comment]
Thanks for reporting this. We're nearing the release of 0.3.9 (currently testing 0.3.9-rc1, but we're going to make sure a fix for this makes it into our 0.3.9-rc2 release candidate).

**Comment by conorsch** [comment]
> This is `grep` with one argument, which causes `grep` to hang waiting for STDIN.

The omission of target file is made on several lines. Looking at all the `grep` calls in the script:

```
$ grep --line-number grep tails_files/install.sh 
89:    if grep -q 'Sandbox 1' /etc/tor/torrc && ! grep -q 'Sandbox 1' /etc/tor/torrc.bak; then
217:  elif grep -q -P '^HidServAuth [a-z2-7]{16}\.onion [A-Za-z0-9+/.]{22}' /etc/tor/torrc ; then
218:    app_document_aths="$(grep ^HidServAuth /etc/tor/torrc | head -n 1 | awk '{ print $2 }')"
222:    app_document_aths="$(grep ^Exec=/usr/local/bin/tor-browser | awk '{ print $2 }')"
242:  elif grep -q -P '^Exec=/usr/local/bin/tor-browser\s+[a-z2-7]{16}\.onion' "${amnesia_desktop}/source.desktop" ; then
243:    app_source_ths="$(grep ^Exec=/usr/local/bin/tor-browser | awk '{ print $2 }')"
283:  if ! grep -q 'ansible' "${tails_live_persistence}/live-additional-software.conf"; then
294:  if ! grep -v "^#.*onion" "${securedrop_ansible_base}/inventory" | grep -q onion; then
314:    if grep -q -P '^HidServAuth [a-z2-7]{16}\.onion [A-Za-z0-9+/.]{22}' /etc/tor/torrc ; then
315:      app_document_aths_info="$(grep ^HidServAuth /etc/tor/torrc | head -n 1)"
386:  if ! grep -q 'custom-nm-hooks' "${tails_live_persistence}/persistence.conf"; then
```

we see that there are two calls to `grep` that inappropriately omit the target file. Specifically, the problematic lines are:

* [ ] 222
* [ ] 243

Both should be revised to include the appropriate source file for finding the hidden service info. Thanks for the excellent bug report, @ryantate! Echoing @fowlslegs's comment, we'll get this into the next release.

**Comment by fowlslegs** [comment]
Hmm... I wasn't sure if grepping for grep would create an infinite loop leading to a black hole that would destroy the universe or not... I guess not. Anyway, merged #1397, which should fix this, but am also going to manually review the whole script since it seems not to have received adequate attention before the recent rewrite was merged. We need to also figure out a testing story... let's talk about that more in #1380.
